File: js/src/cst-rust.mjs
// Rust ↔ `.lino` CST converter (issue #138).
//
// Token-level lossless converter for Rust source. Produces a `lino-cst.rust.*`
// flat CST: every byte of input is encoded as one of the three CST node kinds.
// The round-trip is byte-faithful: `printRust(parseRust(src)) === src`.
//
// Scope: this converter operates at the token-stream level. It recognises:
//
//   - line comments (`// ...`, `/// ...`, `//! ...`)
//   - block comments (`/* ... */`, possibly nested)
//   - whitespace (spaces, tabs, CR, LF)
//   - shebang on the first line
//   - string literals (`"..."`), byte-strings (`b"..."`), raw strings (`r"..."`,
//     `r#"..."#`, etc.), char literals (`'...'`), byte-char literals.
//   - numeric literals (decimal, hex, oct, bin, with suffixes).
//   - identifiers (including raw identifiers `r#foo` and Unicode XID).
//   - punctuation (every other byte is an operator/punctuator token).
//
// The output CST is a single `lino-cst.rust.source_file` list whose children
// alternate between tokens (significant lexemes) and trivia (whitespace,
// comments). This matches the rust-analyzer rowan layout: trivia is preserved
// at the leaf level, structure is preserved at the leaf level too.
//
// A full semantic CST (one `lino-cst.rust.*` tag per `ra_ap_syntax`
// `SyntaxKind`) is built on top of this token stream — that work is
// orthogonal to round-trip fidelity and is intentionally deferred. The
// token-level converter already satisfies issue #138's requirement that "every
// variable and other name, and even whitespace if needed" survive a round
// trip.

import { list, token, trivia, DIALECTS } from './cst.mjs';

const RUST = DIALECTS.rust;

/**
 * Parse Rust source into a `lino-cst.rust.*` CST.
 *
 * The returned tree is a single `source_file` list. Every byte of `src` is
 * preserved verbatim in the resulting tree; `printRust(parseRust(src)) ===
 * src` for any well-tokenised input.
 *
 * @param {string} src Rust source.
 * @returns {CstNode} CST root.
 */
export function parseRust(src) {
  const children = tokeniseRust(String(src));
  return list(`${RUST}.source_file`, children);
}

/**
 * Print a Rust CST back to source. Inverse of `parseRust`.
 *
 * @param {CstNode} node
 * @returns {string}
 */
export function printRust(node) {
  const out = [];
  emit(node, out);
  return out.join('');
}

function emit(node, out) {
  if (!node) return;
  if (node.kind === 'token' || node.kind === 'trivia') {
    out.push(node.text);
    return;
  }
  if (node.kind === 'list') {
    if (node.open) out.push(node.open);
    for (const child of node.children) emit(child, out);
    if (node.close) out.push(node.close);
  }
}

// ---------- Lexer ----------

const ID_START = /[A-Za-z_]/;
const ID_CONT = /[A-Za-z0-9_]/;
const DIGIT = /[0-9]/;
const HEX = /[0-9A-Fa-f]/;

function tokeniseRust(src) {
  const out = [];
  let i = 0;

  if (src.startsWith('#!') && (src.length < 3 || src[2] !== '[')) {
    // Shebang line — preserved as trivia at the start.
    let j = src.indexOf('\n');
    if (j === -1) j = src.length;
    out.push(trivia(src.substring(0, j), `${RUST}.shebang`));
    i = j;
  }

  while (i < src.length) {
    const c = src[i];

    if (c === ' ' || c === '\t' || c === '\r' || c === '\n') {
      let j = i;
      while (j < src.length && (src[j] === ' ' || src[j] === '\t' || src[j] === '\r' || src[j] === '\n')) j++;
      out.push(trivia(src.substring(i, j), `${RUST}.whitespace`));
      i = j;
      continue;
    }

    if (c === '/' && src[i + 1] === '/') {
      let j = i + 2;
      while (j < src.length && src[j] !== '\n') j++;
      out.push(trivia(src.substring(i, j), `${RUST}.comment.line`));
      i = j;
      continue;
    }

    if (c === '/' && src[i + 1] === '*') {
      const j = scanBlockComment(src, i);
      out.push(trivia(src.substring(i, j), `${RUST}.comment.block`));
      i = j;
      continue;
    }

    if (c === '"') {
      const j = scanString(src, i + 1, '"');
      out.push(token(src.substring(i, j), `${RUST}.string_literal`));
      i = j;
      continue;
    }

    if ((c === 'b' || c === 'r') && (src[i + 1] === '"' || (c === 'r' && src[i + 1] === '#') || (c === 'b' && src[i + 1] === 'r' && (src[i + 2] === '"' || src[i + 2] === '#')))) {
      const j = scanRawOrPrefixedString(src, i);
      if (j > i) {
        out.push(token(src.substring(i, j), `${RUST}.string_literal`));
        i = j;
        continue;
      }
    }

    if (c === "'" ) {
      // char literal or lifetime
      const lifetimeEnd = scanLifetime(src, i);
      if (lifetimeEnd > i + 1) {
        out.push(token(src.substring(i, lifetimeEnd), `${RUST}.lifetime`));
        i = lifetimeEnd;
        continue;
      }
      const j = scanString(src, i + 1, "'");
      out.push(token(src.substring(i, j), `${RUST}.char_literal`));
      i = j;
      continue;
    }

    if (c === 'b' && src[i + 1] === "'") {
      const j = scanString(src, i + 2, "'");
      out.push(token(src.substring(i, j), `${RUST}.byte_literal`));
      i = j;
      continue;
    }

    if (DIGIT.test(c)) {
      const j = scanNumber(src, i);
      out.push(token(src.substring(i, j), `${RUST}.numeric_literal`));
      i = j;
      continue;
    }

    if (c === 'r' && src[i + 1] === '#' && ID_START.test(src[i + 2] || '')) {
      let j = i + 2;
      while (j < src.length && ID_CONT.test(src[j])) j++;
      out.push(token(src.substring(i, j), `${RUST}.raw_ident`));
      i = j;
      continue;
    }

    if (ID_START.test(c) || isUnicodeIdStart(c)) {
      let j = i + 1;
      while (j < src.length && (ID_CONT.test(src[j]) || isUnicodeIdContinue(src[j]))) j++;
      out.push(token(src.substring(i, j), `${RUST}.ident`));
      i = j;
      continue;
    }

    // Punctuation: emit one byte. Operators are tokens too; the printer just
    // concatenates them, so we do not need to bundle multi-char operators.
    out.push(token(c, `${RUST}.punct`));
    i += 1;
  }

  return out;
}

function scanBlockComment(src, i) {
  let j = i + 2;
  let depth = 1;
  while (j < src.length && depth > 0) {
    if (src[j] === '/' && src[j + 1] === '*') {
      depth++;
      j += 2;
    } else if (src[j] === '*' && src[j + 1] === '/') {
      depth--;
      j += 2;
    } else {
      j++;
    }
  }
  return j;
}

function scanString(src, j, quote) {
  while (j < src.length) {
    const c = src[j];
    if (c === '\\') { j += 2; continue; }
    if (c === quote) return j + 1;
    j++;
  }
  return j;
}

function scanRawOrPrefixedString(src, i) {
  let j = i;
  if (src[j] === 'b') j++;
  if (src[j] === 'r') {
    j++;
    let hashes = 0;
    while (src[j] === '#') { hashes++; j++; }
    if (src[j] !== '"') return i;
    j++;
    const terminator = '"' + '#'.repeat(hashes);
    const end = src.indexOf(terminator, j);
    if (end === -1) return src.length;
    return end + terminator.length;
  }
  if (src[j] === '"') {
    return scanString(src, j + 1, '"');
  }
  return i;
}

function scanLifetime(src, i) {
  let j = i + 1;
  if (j < src.length && (ID_START.test(src[j]) || isUnicodeIdStart(src[j]))) {
    j++;
    while (j < src.length && (ID_CONT.test(src[j]) || isUnicodeIdContinue(src[j]))) j++;
    // If next char is a quote, this is a char literal, not a lifetime.
    if (src[j] === "'") return i;
    return j;
  }
  return i;
}

function scanNumber(src, i) {
  let j = i;
  if (src[j] === '0' && (src[j + 1] === 'x' || src[j + 1] === 'X')) {
    j += 2;
    while (j < src.length && (HEX.test(src[j]) || src[j] === '_')) j++;
  } else if (src[j] === '0' && (src[j + 1] === 'o' || src[j + 1] === 'O')) {
    j += 2;
    while (j < src.length && (/[0-7_]/.test(src[j]))) j++;
  } else if (src[j] === '0' && (src[j + 1] === 'b' || src[j + 1] === 'B')) {
    j += 2;
    while (j < src.length && (/[01_]/.test(src[j]))) j++;
  } else {
    while (j < src.length && (DIGIT.test(src[j]) || src[j] === '_')) j++;
    if (src[j] === '.' && DIGIT.test(src[j + 1])) {
      j++;
      while (j < src.length && (DIGIT.test(src[j]) || src[j] === '_')) j++;
    }
    if (src[j] === 'e' || src[j] === 'E') {
      j++;
      if (src[j] === '+' || src[j] === '-') j++;
      while (j < src.length && (DIGIT.test(src[j]) || src[j] === '_')) j++;
    }
  }
  // Type suffix: i32, u64, f64, isize, usize, etc.
  if (j < src.length && ID_START.test(src[j])) {
    while (j < src.length && ID_CONT.test(src[j])) j++;
  }
  return j;
}

function isUnicodeIdStart(c) {
  if (!c) return false;
  const code = c.codePointAt(0);
  return code > 0x7f;
}

function isUnicodeIdContinue(c) {
  return isUnicodeIdStart(c);
}

File: rust/src/cst_rust.rs
//! Rust ↔ `.lino` CST converter (issue #138).
//!
//! Token-level lossless converter for Rust source. Produces a
//! `lino-cst.rust.*` flat CST whose round-trip is byte-faithful:
//! `print_rust(&parse_rust(src)) == src`. Mirrors `js/src/cst-rust.mjs`
//! line for line.

use crate::cst::{dialects::RUST, print_cst, CstNode};

/// Parse Rust source into a `lino-cst.rust.*` CST.
pub fn parse_rust(src: &str) -> CstNode {
    let children = tokenise(src);
    CstNode::list(format!("{}.source_file", RUST), children)
}

/// Print a Rust CST back to source.
pub fn print_rust(node: &CstNode) -> String {
    print_cst(node)
}

fn tokenise(src: &str) -> Vec<CstNode> {
    let chars: Vec<char> = src.chars().collect();
    let mut out: Vec<CstNode> = Vec::new();
    let mut i = 0usize;

    if chars.len() >= 2 && chars[0] == '#' && chars[1] == '!' && chars.get(2) != Some(&'[') {
        let mut j = i;
        while j < chars.len() && chars[j] != '\n' {
            j += 1;
        }
        out.push(CstNode::trivia(
            chars[i..j].iter().collect::<String>(),
            Some(&format!("{}.shebang", RUST)),
        ));
        i = j;
    }

    while i < chars.len() {
        let c = chars[i];

        if c == ' ' || c == '\t' || c == '\r' || c == '\n' {
            let mut j = i;
            while j < chars.len()
                && (chars[j] == ' ' || chars[j] == '\t' || chars[j] == '\r' || chars[j] == '\n')
            {
                j += 1;
            }
            out.push(CstNode::trivia(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.whitespace", RUST)),
            ));
            i = j;
            continue;
        }

        if c == '/' && chars.get(i + 1) == Some(&'/') {
            let mut j = i + 2;
            while j < chars.len() && chars[j] != '\n' {
                j += 1;
            }
            out.push(CstNode::trivia(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.comment.line", RUST)),
            ));
            i = j;
            continue;
        }

        if c == '/' && chars.get(i + 1) == Some(&'*') {
            let j = scan_block_comment(&chars, i);
            out.push(CstNode::trivia(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.comment.block", RUST)),
            ));
            i = j;
            continue;
        }

        if c == '"' {
            let j = scan_string(&chars, i + 1, '"');
            out.push(CstNode::token(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.string_literal", RUST)),
            ));
            i = j;
            continue;
        }

        // Raw/byte strings: b"...", r"...", r#"..."#, br"...".
        if (c == 'b' || c == 'r')
            && (chars.get(i + 1) == Some(&'"')
                || (c == 'r' && chars.get(i + 1) == Some(&'#'))
                || (c == 'b'
                    && chars.get(i + 1) == Some(&'r')
                    && (chars.get(i + 2) == Some(&'"') || chars.get(i + 2) == Some(&'#'))))
        {
            if let Some(j) = scan_raw_or_prefixed_string(&chars, i) {
                out.push(CstNode::token(
                    chars[i..j].iter().collect::<String>(),
                    Some(&format!("{}.string_literal", RUST)),
                ));
                i = j;
                continue;
            }
        }

        if c == '\'' {
            let lifetime_end = scan_lifetime(&chars, i);
            if lifetime_end > i + 1 {
                out.push(CstNode::token(
                    chars[i..lifetime_end].iter().collect::<String>(),
                    Some(&format!("{}.lifetime", RUST)),
                ));
                i = lifetime_end;
                continue;
            }
            let j = scan_string(&chars, i + 1, '\'');
            out.push(CstNode::token(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.char_literal", RUST)),
            ));
            i = j;
            continue;
        }

        if c == 'b' && chars.get(i + 1) == Some(&'\'') {
            let j = scan_string(&chars, i + 2, '\'');
            out.push(CstNode::token(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.byte_literal", RUST)),
            ));
            i = j;
            continue;
        }

        if c.is_ascii_digit() {
            let j = scan_number(&chars, i);
            out.push(CstNode::token(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.numeric_literal", RUST)),
            ));
            i = j;
            continue;
        }

        if c == 'r'
            && chars.get(i + 1) == Some(&'#')
            && chars
                .get(i + 2)
                .map(|c| is_ident_start(*c))
                .unwrap_or(false)
        {
            let mut j = i + 2;
            while j < chars.len() && is_ident_continue(chars[j]) {
                j += 1;
            }
            out.push(CstNode::token(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.raw_ident", RUST)),
            ));
            i = j;
            continue;
        }

        if is_ident_start(c) {
            let mut j = i + 1;
            while j < chars.len() && is_ident_continue(chars[j]) {
                j += 1;
            }
            out.push(CstNode::token(
                chars[i..j].iter().collect::<String>(),
                Some(&format!("{}.ident", RUST)),
            ));
            i = j;
            continue;
        }

        out.push(CstNode::token(
            c.to_string(),
            Some(&format!("{}.punct", RUST)),
        ));
        i += 1;
    }

    out
}

fn scan_block_comment(chars: &[char], i: usize) -> usize {
    let mut j = i + 2;
    let mut depth = 1;
    while j < chars.len() && depth > 0 {
        if chars[j] == '/' && chars.get(j + 1) == Some(&'*') {
            depth += 1;
            j += 2;
        } else if chars[j] == '*' && chars.get(j + 1) == Some(&'/') {
            depth -= 1;
            j += 2;
        } else {
            j += 1;
        }
    }
    j
}

fn scan_string(chars: &[char], mut j: usize, quote: char) -> usize {
    while j < chars.len() {
        let c = chars[j];
        if c == '\\' {
            j += 2;
            continue;
        }
        if c == quote {
            return j + 1;
        }
        j += 1;
    }
    j
}

fn scan_raw_or_prefixed_string(chars: &[char], i: usize) -> Option<usize> {
    let mut j = i;
    if chars.get(j) == Some(&'b') {
        j += 1;
    }
    if chars.get(j) == Some(&'r') {
        j += 1;
        let mut hashes = 0;
        while chars.get(j) == Some(&'#') {
            hashes += 1;
            j += 1;
        }
        if chars.get(j) != Some(&'"') {
            return None;
        }
        j += 1;
        let terminator: String =
            std::iter::once('"').chain(std::iter::repeat('#').take(hashes)).collect();
        // search for terminator
        let rest: String = chars[j..].iter().collect();
        match rest.find(&terminator) {
            Some(rel) => Some(j + rel + terminator.chars().count()),
            None => Some(chars.len()),
        }
    } else if chars.get(j) == Some(&'"') {
        Some(scan_string(chars, j + 1, '"'))
    } else {
        None
    }
}

fn scan_lifetime(chars: &[char], i: usize) -> usize {
    let mut j = i + 1;
    if j < chars.len() && is_ident_start(chars[j]) {
        j += 1;
        while j < chars.len() && is_ident_continue(chars[j]) {
            j += 1;
        }
        if chars.get(j) == Some(&'\'') {
            return i;
        }
        return j;
    }
    i
}

fn scan_number(chars: &[char], i: usize) -> usize {
    let mut j = i;
    if chars.get(j) == Some(&'0') && matches!(chars.get(j + 1), Some('x') | Some('X')) {
        j += 2;
        while j < chars.len() && (chars[j].is_ascii_hexdigit() || chars[j] == '_') {
            j += 1;
        }
    } else if chars.get(j) == Some(&'0') && matches!(chars.get(j + 1), Some('o') | Some('O')) {
        j += 2;
        while j < chars.len() && matches!(chars[j], '0'..='7' | '_') {
            j += 1;
        }
    } else if chars.get(j) == Some(&'0') && matches!(chars.get(j + 1), Some('b') | Some('B')) {
        j += 2;
        while j < chars.len() && matches!(chars[j], '0' | '1' | '_') {
            j += 1;
        }
    } else {
        while j < chars.len() && (chars[j].is_ascii_digit() || chars[j] == '_') {
            j += 1;
        }
        if chars.get(j) == Some(&'.')
            && chars.get(j + 1).map(|c| c.is_ascii_digit()).unwrap_or(false)
        {
            j += 1;
            while j < chars.len() && (chars[j].is_ascii_digit() || chars[j] == '_') {
                j += 1;
            }
        }
        if matches!(chars.get(j), Some('e') | Some('E')) {
            j += 1;
            if matches!(chars.get(j), Some('+') | Some('-')) {
                j += 1;
            }
            while j < chars.len() && (chars[j].is_ascii_digit() || chars[j] == '_') {
                j += 1;
            }
        }
    }
    if j < chars.len() && is_ident_start(chars[j]) {
        while j < chars.len() && is_ident_continue(chars[j]) {
            j += 1;
        }
    }
    j
}

fn is_ident_start(c: char) -> bool {
    c == '_' || c.is_ascii_alphabetic() || (c as u32) > 0x7F
}

fn is_ident_continue(c: char) -> bool {
    c == '_' || c.is_ascii_alphanumeric() || (c as u32) > 0x7F
}

#[cfg(test)]
mod tests {
    use super::*;

    fn rt(src: &str) {
        let node = parse_rust(src);
        let back = print_rust(&node);
        assert_eq!(back, src, "round-trip mismatch");
    }

    #[test]
    fn empty_string() {
        rt("");
    }

    #[test]
    fn simple_fn() {
        rt("fn main() {}\n");
    }

    #[test]
    fn line_and_block_comments() {
        rt("// hi\nfn f() {\n    /* mid */ 1\n}\n");
    }

    #[test]
    fn raw_and_byte_strings() {
        rt("let s = r#\"raw\"#;\nlet b = b\"abc\";\n");
    }

    #[test]
    fn lifetime_vs_char() {
        rt("let c = 'a';\nlet lt: &'static str = \"x\";\n");
    }

    #[test]
    fn numeric_with_suffix() {
        rt("let n = 0xFF_FFu32;\nlet f = 3.14e10_f64;\n");
    }

    #[test]
    fn raw_ident() {
        rt("let r#match = 1;\n");
    }
}
